File: python/watermark_cleaner/layers/office.py
import io
import re
import struct
import zipfile
import zlib
from functools import lru_cache
from pathlib import Path

from ..findings import Finding, Report
from ..safeio import is_symlink, write_bytes_atomic

_DOCX_RULES = (
    {"part": "docProps/core.xml", "remove": ("creator", "lastModifiedBy")},
    {"part": "docProps/app.xml", "remove": ("Application", "AppVersion", "Company", "Manager", "Template", "TotalTime")},
    {"part": "word/people.xml", "remove": ("person",)},
    {"part": "word/*.xml", "blank_attributes": ("author", "initials")},
)
_ODT_RULES = (
    {"part": "meta.xml", "remove": ("creator", "initial-creator", "generator", "editing-duration", "printed-by")},
    {"part": "content.xml", "empty": ("creator",)},
)
_RULES = {".docx": _DOCX_RULES, ".odt": _ODT_RULES}

_SUPPORTED_SUFFIXES = tuple(_RULES)

_LOCAL_HEADER_SIZE = 30
_LOCAL_HEADER_SIGNATURE = b"PK\x03\x04"
_CENTRAL_SIGNATURE = b"PK\x01\x02"
_EOCD_SIGNATURE = b"PK\x05\x06"
_STREAMED_FLAG = 0x0008
_ENCRYPTED_FLAG = 0x0001
_MAX_METADATA_BYTES = 64 * 1024 * 1024


def _part_matches(pattern, name):
    if pattern.endswith("/*.xml"):
        directory = pattern[: -len("*.xml")]
        remainder = name[len(directory):]
        return name.startswith(directory) and remainder.endswith(".xml") and "/" not in remainder
    return pattern == name


def _rules_for(rules, name):
    return [rule for rule in rules if _part_matches(rule["part"], name)]


@lru_cache(maxsize=None)
def _remove_pattern(tag):
    name = re.escape(tag)
    return re.compile(
        rf"<([\w.]+:)?{name}\b[^>]*/>|<([\w.]+:)?{name}\b[^>]*>.*?</([\w.]+:)?{name}>",
        re.DOTALL,
    )


@lru_cache(maxsize=None)
def _empty_pattern(tag):
    name = re.escape(tag)
    return re.compile(rf"(<(?:[\w.]+:)?{name}\b[^>]*>)(.*?)(</(?:[\w.]+:)?{name}>)", re.DOTALL)


@lru_cache(maxsize=None)
def _attribute_pattern(attribute):
    name = re.escape(attribute)
    return re.compile(rf"(\s(?:[\w.]+:)?{name}=)([\"'])([^\"']*)\2")


def _scrub(text, rule):
    count = 0
    for tag in rule.get("remove", ()):
        text, n = _remove_pattern(tag).subn("", text)
        count += n
    for tag in rule.get("empty", ()):
        pattern = _empty_pattern(tag)
        count += sum(1 for match in pattern.finditer(text) if match.group(2))
        text = pattern.sub(_keep_open_and_close, text)
    for attribute in rule.get("blank_attributes", ()):
        pattern = _attribute_pattern(attribute)
        count += sum(1 for match in pattern.finditer(text) if match.group(3))
        text = pattern.sub(_blank_attribute_value, text)
    return text, count


def _keep_open_and_close(match):
    return match.group(1) + match.group(3)


def _blank_attribute_value(match):
    return match.group(1) + match.group(2) + match.group(2)


def inspect_file(path):
    return _process(path, write=False, backup=False)


def clean_file(path, write=True, backup=True):
    return _process(path, write=write, backup=backup)


def _process(path, write, backup):
    suffix = Path(path).suffix.lower()
    report = Report(path=str(path))
    rules = _RULES.get(suffix)
    if rules is None:
        return report
    data = Path(path).read_bytes()
    result = _clean_zip(data, rules)
    if result is None:
        report.findings.append(Finding("metadata", "office", "warn", "could not parse (file left untouched)", 1))
        return report
    new_data, stripped = result
    if not stripped:
        return report
    if not write:
        report.findings.append(Finding("metadata", "office", "warn", "embedded author/application metadata present", stripped))
        return report
    if is_symlink(path):
        report.findings.append(Finding("io", "write", "warn", "refused to write through symlink", 1))
        return report
    report.changed = True
    if backup:
        _backup_file(path)
    write_bytes_atomic(path, new_data)
    report.findings.append(
        Finding("metadata", "office", "fixed", "stripped author/application metadata (lossless, container-level)", stripped)
    )
    return report


def _clean_zip(data, rules):
    try:
        zf = zipfile.ZipFile(io.BytesIO(data))
        infos = zf.infolist()
    except (zipfile.BadZipFile, OSError, NotImplementedError, ValueError, RuntimeError):
        return None

    records = []
    stripped_total = 0
    for info in infos:
        if info.compress_type not in (zipfile.ZIP_STORED, zipfile.ZIP_DEFLATED):
            return None
        if info.flag_bits & _ENCRYPTED_FLAG:
            return None
        if info.compress_size > 0xFFFFFFFF or info.file_size > 0xFFFFFFFF or info.header_offset > 0xFFFFFFFF:
            return None
        span = _local_data_span(data, info)
        if span is None:
            return None
        header_bytes, comp_data = span
        crc, csize, usize, method = info.CRC, info.compress_size, info.file_size, info.compress_type
        modified = False
        part_rules = _rules_for(rules, info.filename)
        if part_rules:
            if info.file_size > _MAX_METADATA_BYTES:
                return None
            content = _decompress(comp_data, info.compress_type)
            if content is None:
                return None
            try:
                text = content.decode("utf-8")
            except UnicodeDecodeError:
                text = None
            if text is not None:
                new_text = text
                n = 0
                for rule in part_rules:
                    new_text, rule_count = _scrub(new_text, rule)
                    n += rule_count
                if n:
                    stripped_total += n
                    modified = True
                    new_bytes = new_text.encode("utf-8")
                    comp_data = new_bytes
                    method = zipfile.ZIP_STORED
                    crc = zlib.crc32(new_bytes) & 0xFFFFFFFF
                    csize = len(comp_data)
                    usize = len(new_bytes)
        records.append((info, header_bytes, comp_data, crc, csize, usize, method, modified))

    if not stripped_total:
        return data, 0
    return _write_zip(records), stripped_total


def _local_data_span(data, info):
    offset = info.header_offset
    if offset < 0 or offset + _LOCAL_HEADER_SIZE > len(data):
        return None
    header = data[offset : offset + _LOCAL_HEADER_SIZE]
    if header[0:4] != _LOCAL_HEADER_SIGNATURE:
        return None
    name_len, extra_len = struct.unpack("<HH", header[26:30])
    data_start = offset + _LOCAL_HEADER_SIZE + name_len + extra_len
    data_end = data_start + info.compress_size
    if data_end > len(data):
        return None
    return data[offset:data_start], data[data_start:data_end]


def _decompress(data, method):
    if method == zipfile.ZIP_STORED:
        return data
    try:
        obj = zlib.decompressobj(-15)
        out = obj.decompress(data, _MAX_METADATA_BYTES + 1)
    except zlib.error:
        return None
    if len(out) > _MAX_METADATA_BYTES or obj.unconsumed_tail:
        return None
    return out


def _patch_header(header_bytes, crc, csize, usize, method):
    header = bytearray(header_bytes)
    flags = struct.unpack("<H", header[6:8])[0] & ~_STREAMED_FLAG
    header[6:8] = struct.pack("<H", flags)
    header[8:10] = struct.pack("<H", method)
    header[14:18] = struct.pack("<I", crc)
    header[18:22] = struct.pack("<I", csize)
    header[22:26] = struct.pack("<I", usize)
    return bytes(header)


def _dos_datetime(date_time):
    year, month, day, hour, minute, second = date_time
    dosdate = ((max(year, 1980) - 1980) << 9) | (month << 5) | day
    dostime = (hour << 11) | (minute << 5) | (second // 2)
    return dostime, dosdate


def _central_record(info, crc, csize, usize, method, local_offset):
    name = info.filename.encode("utf-8")
    extra = info.extra or b""
    comment = info.comment or b""
    dostime, dosdate = _dos_datetime(info.date_time)
    flags = info.flag_bits & ~_STREAMED_FLAG
    version_made_by = (info.create_system << 8) | info.create_version
    return (
        _CENTRAL_SIGNATURE
        + struct.pack(
            "<HHHHHHIIIHHHHHII",
            version_made_by,
            info.extract_version,
            flags,
            method,
            dostime,
            dosdate,
            crc,
            csize,
            usize,
            len(name),
            len(extra),
            len(comment),
            0,
            info.internal_attr,
            info.external_attr,
            local_offset,
        )
        + name
        + extra
        + comment
    )


def _eocd(count, cd_size, cd_offset):
    return _EOCD_SIGNATURE + struct.pack("<HHHHIIH", 0, 0, count, count, cd_size, cd_offset, 0)


def _write_zip(records):
    ordered = sorted(records, key=lambda r: r[0].header_offset)
    out = bytearray()
    offsets = []
    for info, header_bytes, comp_data, crc, csize, usize, method, modified in ordered:
        offsets.append(len(out))
        out += _patch_header(header_bytes, crc, csize, usize, method) if modified else header_bytes
        out += comp_data
    central = bytearray()
    for (info, header_bytes, comp_data, crc, csize, usize, method, modified), offset in zip(ordered, offsets):
        central += _central_record(info, crc, csize, usize, method, offset)
    cd_offset = len(out)
    out += central
    out += _eocd(len(ordered), len(central), cd_offset)
    return bytes(out)


def _backup_file(path):
    p = Path(path)
    backup = p.with_suffix(p.suffix + ".bak")
    if not backup.exists():
        backup.write_bytes(p.read_bytes())

File: node/src/office.js
const fs = require("fs");
const path = require("path");
const zlib = require("zlib");

const { isSymlink, writeAtomic } = require("./safeio");

const DOCX_CORE_TAGS = ["creator", "lastModifiedBy"];
const DOCX_APP_TAGS = ["Application", "Company", "Manager"];
const ODT_META_TAGS = ["creator", "initial-creator", "generator"];

const TARGETS = {
  ".docx": { "docProps/core.xml": DOCX_CORE_TAGS, "docProps/app.xml": DOCX_APP_TAGS },
  ".odt": { "meta.xml": ODT_META_TAGS },
};

const ALL_TAGS = new Set();
for (const targets of Object.values(TARGETS)) {
  for (const tags of Object.values(targets)) {
    for (const tag of tags) ALL_TAGS.add(tag);
  }
}

function tagPattern(tag) {
  const name = tag.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`<([\\w.]+:)?${name}\\b[^>]*/>|<([\\w.]+:)?${name}\\b[^>]*>[\\s\\S]*?</([\\w.]+:)?${name}>`, "g");
}

const PATTERNS = {};
for (const tag of ALL_TAGS) PATTERNS[tag] = tagPattern(tag);

function stripTags(text, tags) {
  let count = 0;
  for (const tag of tags) {
    text = text.replace(PATTERNS[tag], () => {
      count += 1;
      return "";
    });
  }
  return { text, count };
}

function emptyReport(file) {
  return { path: file, findings: [], changed: false, counts: { fixed: 0, warn: 0, error: 0 }, has_errors: false };
}

function addFinding(report, severity, kind, message, count) {
  report.findings.push({ layer: "metadata", kind, severity, message, count, examples: [] });
  report.counts[severity] = (report.counts[severity] || 0) + count;
}

function inspectFile(file) {
  return processFile(file, false, false);
}

function cleanFile(file, write, backup) {
  return processFile(file, write, backup);
}

function processFile(file, write, backup) {
  const suffix = path.extname(file).toLowerCase();
  const report = emptyReport(file);
  const targets = TARGETS[suffix];
  if (!targets) return report;
  const data = fs.readFileSync(file);
  const result = cleanZip(data, targets);
  if (result === null) {
    addFinding(report, "warn", "office", "could not parse (file left untouched)", 1);
    return report;
  }
  const { data: newData, stripped } = result;
  if (!stripped) return report;
  if (!write) {
    addFinding(report, "warn", "office", "embedded author/application metadata present", stripped);
    return report;
  }
  if (isSymlink(file)) {
    addFinding(report, "warn", "write", "refused to write through symlink", 1);
    return report;
  }
  report.changed = true;
  if (backup) backupFile(file);
  writeAtomic(file, newData);
  addFinding(report, "fixed", "office", "stripped author/application metadata (lossless, container-level)", stripped);
  return report;
}

const EOCD_SIGNATURE = 0x06054b50;
const CENTRAL_SIGNATURE = 0x02014b50;
const LOCAL_SIGNATURE = 0x04034b50;
const STREAMED_FLAG = 0x0008;
const ENCRYPTED_FLAG = 0x0001;
const ZIP_STORED = 0;
const ZIP_DEFLATED = 8;
const MAX_METADATA_BYTES = 64 * 1024 * 1024;

const strictUtf8 = new TextDecoder("utf-8", { fatal: true });

function findEocd(data) {
  const maxCommentLen = 65535;
  const minPos = Math.max(0, data.length - 22 - maxCommentLen);
  for (let i = data.length - 22; i >= minPos; i -= 1) {
    if (data.readUInt32LE(i) === EOCD_SIGNATURE) {
      const commentLen = data.readUInt16LE(i + 20);
      if (i + 22 + commentLen === data.length) return i;
    }
  }
  return -1;
}

function parseCentralDirectory(data) {
  if (data.length < 22) return null;
  const eocdOffset = findEocd(data);
  if (eocdOffset === -1) return null;
  const totalEntries = data.readUInt16LE(eocdOffset + 10);
  const cdSize = data.readUInt32LE(eocdOffset + 12);
  const cdOffset = data.readUInt32LE(eocdOffset + 16);
  if (totalEntries === 0xffff || cdOffset === 0xffffffff || cdSize === 0xffffffff) return null;
  if (cdOffset + cdSize > data.length) return null;

  const entries = [];
  let pos = cdOffset;
  for (let i = 0; i < totalEntries; i += 1) {
    if (pos + 46 > data.length) return null;
    if (data.readUInt32LE(pos) !== CENTRAL_SIGNATURE) return null;
    const versionMadeBy = data.readUInt16LE(pos + 4);
    const versionNeeded = data.readUInt16LE(pos + 6);
    const flags = data.readUInt16LE(pos + 8);
    const method = data.readUInt16LE(pos + 10);
    const modTime = data.readUInt16LE(pos + 12);
    const modDate = data.readUInt16LE(pos + 14);
    const crc = data.readUInt32LE(pos + 16);
    const csize = data.readUInt32LE(pos + 20);
    const usize = data.readUInt32LE(pos + 24);
    const nameLen = data.readUInt16LE(pos + 28);
    const extraLen = data.readUInt16LE(pos + 30);
    const commentLen = data.readUInt16LE(pos + 32);
    const internalAttr = data.readUInt16LE(pos + 36);
    const externalAttr = data.readUInt32LE(pos + 38);
    const localOffset = data.readUInt32LE(pos + 42);
    if (csize === 0xffffffff || usize === 0xffffffff || localOffset === 0xffffffff) return null;
    const nameStart = pos + 46;
    const nameEnd = nameStart + nameLen;
    const extraEnd = nameEnd + extraLen;
    const commentEnd = extraEnd + commentLen;
    if (commentEnd > data.length) return null;
    entries.push({
      filename: data.subarray(nameStart, nameEnd).toString("utf-8"),
      versionMadeBy,
      versionNeeded,
      flags,
      method,
      modTime,
      modDate,
      crc,
      csize,
      usize,
      internalAttr,
      externalAttr,
      localOffset,
      extra: Buffer.from(data.subarray(nameEnd, extraEnd)),
      comment: Buffer.from(data.subarray(extraEnd, commentEnd)),
    });
    pos = commentEnd;
  }
  if (pos !== cdOffset + cdSize) return null;
  return entries;
}

function localDataSpan(data, entry) {
  const offset = entry.localOffset;
  if (offset < 0 || offset + 30 > data.length) return null;
  if (data.readUInt32LE(offset) !== LOCAL_SIGNATURE) return null;
  const nameLen = data.readUInt16LE(offset + 26);
  const extraLen = data.readUInt16LE(offset + 28);
  const dataStart = offset + 30 + nameLen + extraLen;
  const dataEnd = dataStart + entry.csize;
  if (dataEnd > data.length) return null;
  return { header: Buffer.from(data.subarray(offset, dataStart)), compData: data.subarray(dataStart, dataEnd) };
}

function decompressEntry(compData, method) {
  if (method === ZIP_STORED) {
    return compData.length > MAX_METADATA_BYTES ? null : compData;
  }
  try {
    return zlib.inflateRawSync(compData, { maxOutputLength: MAX_METADATA_BYTES });
  } catch (error) {
    return null;
  }
}

function patchHeader(header, crc, csize, usize, method) {
  const patched = Buffer.from(header);
  let flags = patched.readUInt16LE(6);
  flags &= ~STREAMED_FLAG;
  patched.writeUInt16LE(flags, 6);
  patched.writeUInt16LE(method, 8);
  patched.writeUInt32LE(crc, 14);
  patched.writeUInt32LE(csize, 18);
  patched.writeUInt32LE(usize, 22);
  return patched;
}

const CRC_TABLE = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n += 1) {
    let c = n;
    for (let k = 0; k < 8; k += 1) {
      c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c >>> 0;
  }
  return table;
})();

function crc32(buffer) {
  let crc = 0xffffffff;
  for (let i = 0; i < buffer.length; i += 1) {
    crc = CRC_TABLE[(crc ^ buffer[i]) & 0xff] ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function cleanZip(data, targets) {
  const entries = parseCentralDirectory(data);
  if (entries === null) return null;

  const records = [];
  let strippedTotal = 0;
  for (const entry of entries) {
    if (entry.method !== ZIP_STORED && entry.method !== ZIP_DEFLATED) return null;
    if (entry.flags & ENCRYPTED_FLAG) return null;
    const span = localDataSpan(data, entry);
    if (span === null) return null;
    let compData = span.compData;
    let crc = entry.crc;
    let csize = entry.csize;
    let usize = entry.usize;
    let method = entry.method;
    let modified = false;
    const tags = targets[entry.filename];
    if (tags) {
      if (entry.usize > MAX_METADATA_BYTES) return null;
      const content = decompressEntry(compData, entry.method);
      if (content === null) return null;
      let text = null;
      try {
        text = strictUtf8.decode(content);
      } catch (error) {
        text = null;
      }
      if (text !== null) {
        const { text: newText, count } = stripTags(text, tags);
        if (count) {
          strippedTotal += count;
          modified = true;
          const newBytes = Buffer.from(newText, "utf-8");
          compData = newBytes;
          method = ZIP_STORED;
          crc = crc32(newBytes);
          csize = compData.length;
          usize = newBytes.length;
        }
      }
    }
    records.push({ entry, header: span.header, compData, crc, csize, usize, method, modified });
  }

  if (!strippedTotal) return { data, stripped: 0 };
  return { data: writeZip(records), stripped: strippedTotal };
}

function centralRecord(entry, crc, csize, usize, method, localOffset) {
  const name = Buffer.from(entry.filename, "utf-8");
  const flags = entry.flags & ~STREAMED_FLAG;
  const header = Buffer.alloc(46);
  header.writeUInt32LE(CENTRAL_SIGNATURE, 0);
  header.writeUInt16LE(entry.versionMadeBy, 4);
  header.writeUInt16LE(entry.versionNeeded, 6);
  header.writeUInt16LE(flags, 8);
  header.writeUInt16LE(method, 10);
  header.writeUInt16LE(entry.modTime, 12);
  header.writeUInt16LE(entry.modDate, 14);
  header.writeUInt32LE(crc, 16);
  header.writeUInt32LE(csize, 20);
  header.writeUInt32LE(usize, 24);
  header.writeUInt16LE(name.length, 28);
  header.writeUInt16LE(entry.extra.length, 30);
  header.writeUInt16LE(entry.comment.length, 32);
  header.writeUInt16LE(0, 34);
  header.writeUInt16LE(entry.internalAttr, 36);
  header.writeUInt32LE(entry.externalAttr, 38);
  header.writeUInt32LE(localOffset, 42);
  return Buffer.concat([header, name, entry.extra, entry.comment]);
}

function buildEocd(count, cdSize, cdOffset) {
  const buf = Buffer.alloc(22);
  buf.writeUInt32LE(EOCD_SIGNATURE, 0);
  buf.writeUInt16LE(0, 4);
  buf.writeUInt16LE(0, 6);
  buf.writeUInt16LE(count, 8);
  buf.writeUInt16LE(count, 10);
  buf.writeUInt32LE(cdSize, 12);
  buf.writeUInt32LE(cdOffset, 16);
  buf.writeUInt16LE(0, 20);
  return buf;
}

function writeZip(records) {
  const ordered = [...records].sort((a, b) => a.entry.localOffset - b.entry.localOffset);
  const localParts = [];
  const offsets = [];
  let pos = 0;
  for (const record of ordered) {
    offsets.push(pos);
    const header = record.modified
      ? patchHeader(record.header, record.crc, record.csize, record.usize, record.method)
      : record.header;
    localParts.push(header, record.compData);
    pos += header.length + record.compData.length;
  }
  const centralParts = ordered.map((record, i) =>
    centralRecord(record.entry, record.crc, record.csize, record.usize, record.method, offsets[i])
  );
  const central = Buffer.concat(centralParts);
  const cdOffset = pos;
  const eocd = buildEocd(ordered.length, central.length, cdOffset);
  return Buffer.concat([...localParts, central, eocd]);
}

function backupFile(file) {
  const bak = `${file}.bak`;
  if (!fs.existsSync(bak)) fs.copyFileSync(file, bak);
}

module.exports = { inspectFile, cleanFile, crc32, TARGETS };
